File: packages/skill/src/fence.ts
import {
  isInsideAnyRoot,
  matchesAnyPattern,
  toolError,
  type ToolError,
} from "@agent-sh/harness-core";
import type {
  LoadedSkill,
  SkillPermissionPolicy,
  SkillSessionConfig,
  SkillTrustMode,
} from "./types.js";

/**
 * Returns a ToolError if the skill dir is out-of-workspace or matches
 * sensitive patterns and no hook is wired.
 */
export function fenceSkill(
  session: SkillSessionConfig,
  skill: LoadedSkill,
): ToolError | undefined {
  const { permissions } = session;
  const isSensitive = matchesAnyPattern(
    skill.dir,
    permissions.sensitivePatterns,
  );
  if (isSensitive && permissions.hook === undefined) {
    return toolError(
      "SENSITIVE",
      `Refusing to activate skill in sensitive path: ${skill.dir}`,
      { meta: { name: skill.name, dir: skill.dir } },
    );
  }
  const inside = isInsideAnyRoot(skill.dir, permissions.roots);
  if (
    !inside &&
    permissions.bypassWorkspaceGuard !== true &&
    permissions.hook === undefined
  ) {
    return toolError(
      "OUTSIDE_WORKSPACE",
      `Skill directory is outside all configured workspace roots: ${skill.dir}`,
      {
        meta: {
          name: skill.name,
          dir: skill.dir,
          roots: permissions.roots,
        },
      },
    );
  }
  return undefined;
}

/**
 * Resolve the trust mode for a skill. See design §5.5. Defaults:
 *
 * - Skills under `session.trust.trustedRoots` always activate (hook
 *   still runs as advisory).
 * - Skills elsewhere fall back to
 *   `session.trust.untrustedProjectSkills` (default `hook_required`).
 */
export function resolveTrustMode(
  session: SkillSessionConfig,
  skillDir: string,
): { trusted: boolean; mode: SkillTrustMode } {
  const trusted = (session.trust?.trustedRoots ?? []).some((root) =>
    isUnderRoot(skillDir, root),
  );
  if (trusted) {
    return { trusted: true, mode: "allow" };
  }
  const mode: SkillTrustMode =
    session.trust?.untrustedProjectSkills ?? "hook_required";
  return { trusted: false, mode };
}

/**
 * Permission hook invocation. Returns `allow`, `allow_once`, or
 * `deny` with a reason. `ask` maps to `deny` in autonomous mode.
 */
export async function askPermission(
  session: SkillSessionConfig,
  args: {
    skill: LoadedSkill;
    reason: "normal" | "untrusted_project_skill";
  },
): Promise<
  | { decision: "allow" | "allow_once" }
  | { decision: "deny"; why: string }
> {
  const perms: SkillPermissionPolicy = session.permissions;
  const pattern = `Skill(name:${args.skill.name})`;
  if (perms.hook === undefined) {
    if (args.reason === "untrusted_project_skill") {
      return {
        decision: "deny",
        why: "untrusted project skill; no permission hook is configured to review it",
      };
    }
    if (perms.unsafeAllowSkillWithoutHook === true) {
      return { decision: "allow" };
    }
    return {
      decision: "deny",
      why: "skill tool has no permission hook configured; refusing to activate untrusted skills. Wire a hook or set session.permissions.unsafeAllowSkillWithoutHook for test fixtures.",
    };
  }
  const decision = await perms.hook({
    tool: "skill",
    path: args.skill.dir,
    action: "activate",
    always_patterns: [pattern],
    metadata: {
      name: args.skill.name,
      root_index: args.skill.rootIndex,
      frontmatter: args.skill.frontmatter,
      reason: args.reason,
    },
  });
  if (decision === "deny") {
    return {
      decision: "deny",
      why: `skill activation blocked by permission policy. Pattern hint: ${pattern}`,
    };
  }
  if (decision === "allow" || decision === "allow_once") {
    return { decision };
  }
  return {
    decision: "deny",
    why: "permission hook returned 'ask' but skill runs in autonomous mode. Configure the hook to return allow or deny.",
  };
}

function isUnderRoot(candidate: string, root: string): boolean {
  if (candidate === root) return true;
  if (!candidate.startsWith(root)) return false;
  const next = candidate.charCodeAt(root.length);
  return next === 0x2f /* '/' */ || next === 0x5c; /* '\\' */
}

File: crates/skill/src/fence.rs
use harness_core::permissions::PermissionQuery;
use harness_core::{PermissionDecision, ToolError, ToolErrorCode};
use std::path::Path;

use crate::types::{LoadedSkill, SkillSessionConfig, SkillTrustMode};

#[derive(Debug, Clone)]
pub enum PermissionOutcome {
    Allow,
    Deny { why: String },
}

pub enum PermissionReason {
    Normal,
    UntrustedProjectSkill,
}

pub fn fence_skill(
    session: &SkillSessionConfig,
    skill: &LoadedSkill,
) -> Option<ToolError> {
    let perms = &session.permissions.inner;
    let path_str = skill.dir.clone();

    let is_sensitive = perms
        .sensitive_patterns
        .iter()
        .any(|p| matches_pattern(&path_str, p));
    if is_sensitive && perms.hook.is_none() {
        return Some(
            ToolError::new(
                ToolErrorCode::Sensitive,
                format!("Refusing to activate skill in sensitive path: {}", path_str),
            )
            .with_meta(serde_json::json!({ "name": skill.name, "dir": path_str })),
        );
    }

    let inside = perms.roots.iter().any(|root| is_inside(&path_str, root));
    if !inside && !perms.bypass_workspace_guard && perms.hook.is_none() {
        return Some(
            ToolError::new(
                ToolErrorCode::OutsideWorkspace,
                format!(
                    "Skill directory is outside all configured workspace roots: {}",
                    path_str
                ),
            )
            .with_meta(serde_json::json!({
                "name": skill.name,
                "dir": path_str,
                "roots": perms.roots,
            })),
        );
    }
    None
}

pub fn resolve_trust_mode(session: &SkillSessionConfig, skill_dir: &str) -> (bool, SkillTrustMode) {
    let trusted = session
        .trust
        .trusted_roots
        .iter()
        .any(|r| is_inside(skill_dir, r));
    if trusted {
        return (true, SkillTrustMode::Allow);
    }
    let mode = session
        .trust
        .untrusted_project_skills
        .unwrap_or(SkillTrustMode::HookRequired);
    (false, mode)
}

pub async fn ask_permission(
    session: &SkillSessionConfig,
    skill: &LoadedSkill,
    reason: PermissionReason,
) -> PermissionOutcome {
    let perms = &session.permissions.inner;
    let pattern = format!("Skill(name:{})", skill.name);

    if perms.hook.is_none() {
        if matches!(reason, PermissionReason::UntrustedProjectSkill) {
            return PermissionOutcome::Deny {
                why: "untrusted project skill; no permission hook is configured to review it"
                    .to_string(),
            };
        }
        if session.permissions.unsafe_allow_skill_without_hook {
            return PermissionOutcome::Allow;
        }
        return PermissionOutcome::Deny {
            why: "skill tool has no permission hook configured; refusing to activate untrusted skills. Wire a hook or set permissions.unsafe_allow_skill_without_hook for test fixtures."
                .to_string(),
        };
    }

    let reason_str = match reason {
        PermissionReason::Normal => "normal",
        PermissionReason::UntrustedProjectSkill => "untrusted_project_skill",
    };
    let metadata = serde_json::json!({
        "name": skill.name,
        "root_index": skill.root_index,
        "frontmatter": skill.frontmatter,
        "reason": reason_str,
    });
    let query = PermissionQuery {
        tool: "skill".to_string(),
        path: skill.dir.clone(),
        action: "activate".to_string(),
        always_patterns: vec![pattern.clone()],
        metadata,
    };
    let hook = perms.hook.as_ref().unwrap();
    let decision = hook(query).await;
    match decision {
        PermissionDecision::Allow | PermissionDecision::AllowOnce => PermissionOutcome::Allow,
        PermissionDecision::Deny => PermissionOutcome::Deny {
            why: format!(
                "skill activation blocked by permission policy. Pattern hint: {}",
                pattern
            ),
        },
        PermissionDecision::Ask => PermissionOutcome::Deny {
            why: "permission hook returned 'ask' but skill runs in autonomous mode. Configure the hook to return allow or deny.".to_string(),
        },
    }
}

fn is_inside(candidate: &str, root: &str) -> bool {
    if candidate == root {
        return true;
    }
    if !candidate.starts_with(root) {
        return false;
    }
    // Only consider it inside if the next char is a path separator.
    match candidate.as_bytes().get(root.len()) {
        Some(b'/') | Some(b'\\') => true,
        None => true,
        _ => false,
    }
}

fn matches_pattern(path: &str, pattern: &str) -> bool {
    // Exact string match first — cheap.
    if path == pattern {
        return true;
    }
    // Use globset for real glob semantics, mirroring the TS matcher at
    // packages/core/src/permissions.ts. `literal_separator(true)` keeps
    // `*` from crossing path segments. Fallback to substring if the
    // pattern is globset-unparseable.
    let pat_lower = pattern.to_ascii_lowercase();
    let path_lower = path.to_ascii_lowercase();
    let base_name = Path::new(&path_lower)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("");
    match globset::GlobBuilder::new(&pat_lower)
        .literal_separator(true)
        .build()
    {
        Ok(glob) => {
            let m = glob.compile_matcher();
            m.is_match(&path_lower) || m.is_match(base_name)
        }
        Err(_) => path_lower.contains(&pat_lower),
    }
}
